src/main/java/hudson/remoting/RemoteClassLoader.java

```diff
@@ -130,10 +133,48 @@ private RemoteClassLoader(@CheckForNull ClassLoader parent, @Nonnull IClassLoade
         final Channel channel = RemoteInvocationHandler.unwrap(proxy);
         this.channel = channel == null ? null : channel.ref();
         this.underlyingProxy = proxy;
+        final IClassLoader targetProxy;
         if (channel == null || !channel.remoteCapability.supportsPrefetch() || channel.getJarCache()==null) {
-            proxy = new DumbClassLoaderBridge(proxy);
+            targetProxy = new DumbClassLoaderBridge(proxy);
+        } else {
+            targetProxy = proxy;
         }
-        this.proxy = proxy;
+        // JENKINS-61103: wrap the target (IClassLoader) proxy in a dynamic proxy which will handle retries in case of
+        // interrupt exception from the remoting system during class loading/initialization.
+        // It should take care of cases like JENKINS-36991, JENKINS-51854 and others.
+        this.proxy = (IClassLoader) Proxy.newProxyInstance(RemoteClassLoader.class.getClassLoader(),
+                new Class[] { IClassLoader.class },
+                (p, method, methodArgs) -> {
+                    Object result;
+                    boolean interrupted = false;
+                    try {
+                        while (true) {
+                            try {
+                                result = method.invoke(targetProxy, methodArgs);
+                                break;
+                            } catch (InvocationTargetException x) {
+                                Throwable e = x.getCause();
+                                if (e instanceof RemotingSystemException
+                                        // this really look like remote interruptions we should defer, to avoid screwing
+                                        // up class loading/initialization:
+                                        && (e.getCause() instanceof InterruptedException
+                                                || e.getCause() instanceof InterruptedIOException)
+                                        // this is only called from PreloadJarTask, we're not actually loading classes:
+                                        && !"fetchJar".equals(method.getName())) {
+                                    interrupted = true;
```

Comment:
This looks like it could get blocked forever. It needs some kind of timeout or retry counter.

Reply:
I can't think of how it could happen for `InterruptedException`, but it probably can for `InterruptedIOException`, so yes, better safe than sorry. 

85a19dee adds a retries counter, with a configurable limit (2 by default; I even think 1 would be enough to handle common scenarios actually).

This also gives a convenient way to disable the retry logic and the whole dynamic proxy thing (set `MAX_RETRIES_ON_REMOTING_INTERRUPTION` to 0), for instance to check that some tests actually fail without it.